File: Solutions/day2/py/main.py

def part_a(arr: list[tuple[str, int]]) -> int:
    horizontal = 0
    depth = 0
    for cmd, num in arr:
        if cmd == "forward":
            horizontal += num
        elif cmd == "up":
            depth -= num
        elif cmd == "down":
            depth += num
    return horizontal * depth


def part_b(arr) -> int:
    horizontal = 0
    depth = 0
    aim = 0
    for cmd, num in arr:
        if cmd == "forward":
            horizontal += num
            depth += (num * aim)
        elif cmd == "up":
            aim -= num
        elif cmd == "down":
            aim += num
    return horizontal * depth


def main() -> tuple[int, int]:
    with open("../input.txt") as f:
        arr = []
        for line in f:
            cmd, num = line.split(" ")
            arr.append((cmd, int(num)))

    return part_a(arr), part_b(arr)


if __name__ == '__main__':
    import time

    start_time = time.perf_counter()
    a, b = main()
    end_time = time.perf_counter()
    print(f"Result part_a: {a}")
    print(f"Result part_b: {b}")
    print(f"Time: {(end_time - start_time):4f} sec")

File: Solutions/day2/rust/src/main.rs
use std::fs;


struct Command {
    cmd: Cmds,
    num: i32
}

enum Cmds {
    Forward,
    Up,
    Down
}

impl Command {

    fn parse(s: &str) -> Command{
        const COMMAND_FORWARD: &str = "forward";
        const COMMAND_UP: &str = "up";
        const COMMAND_DOWN: &str = "down";

        let (cmd, num) = s.split_once(" ").unwrap();
        let cmd = match cmd {
            COMMAND_FORWARD => Cmds::Forward,
            COMMAND_UP      => Cmds::Up,
            COMMAND_DOWN    => Cmds::Down,
            _               => panic!()
        };
        Command {
            cmd,
            num: num.parse().unwrap()
        }
    }
}

fn part_a(v: &[Command]) -> i32{
    let mut horizontal = 0;
    let mut depth = 0;
    for command in v {
        match command.cmd {
            Cmds::Forward => horizontal += command.num,
            Cmds::Up      => depth -= command.num,
            Cmds::Down    => depth += command.num,
        };
    }
    horizontal * depth
}

fn part_b(v: &[Command]) -> i32{
    let mut horizontal = 0;
    let mut depth = 0;
    let mut aim = 0;
    for command in v {
        match command.cmd {
            Cmds::Forward => {
                horizontal += command.num;
                depth += aim*command.num;
            },
            Cmds::Up      => aim -= command.num,
            Cmds::Down    => aim += command.num,
        };
    }
    horizontal * depth
}

fn main() {
    let timer = std::time::Instant::now();
    let input_vec = fs::read_to_string("../input.txt")
        .unwrap()
        .lines()
        .map(|x| Command::parse(x))
        .collect::<Vec<_>>();
    let a = part_a(&input_vec);
    let b = part_b(&input_vec);
    let total_time = timer.elapsed();

    println!("Result part_a: {}", a);
    println!("Result part_b: {}", b);
    println!("Time: {} sec", total_time.as_secs_f32());
}